File: examples/43-advanced-theme-management/main.go
package main

import (
	"fmt"
	"os"
	"path/filepath"

	log "github.com/djinn-soul/gopptx/pkg/stdlog"

	"github.com/djinn-soul/gopptx/pkg/pptx"
	"github.com/djinn-soul/gopptx/pkg/pptx/styling"
)

const (
	outputDir  = "examples/output"
	outputFile = "43_advanced_theme_management.pptx"
)

func main() {
	if err := run(); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
}

func run() error {
	if err := os.MkdirAll(outputDir, 0o750); err != nil {
		return fmt.Errorf("create output dir: %w", err)
	}

	// Step 1: build a base presentation with several content slides.
	tmpDir, err := os.MkdirTemp("", "gopptx-theme-mgmt-*")
	if err != nil {
		return fmt.Errorf("create temp dir: %w", err)
	}
	defer func() { _ = os.RemoveAll(tmpDir) }()

	tmpPath := filepath.Join(tmpDir, "base.pptx")
	if err := buildBase(tmpPath); err != nil {
		return fmt.Errorf("build base: %w", err)
	}

	// Step 2: open the saved file with the editor.
	e, err := pptx.OpenPresentationEditor(tmpPath)
	if err != nil {
		return fmt.Errorf("open editor: %w", err)
	}
	defer func() {
		if closeErr := e.Close(); closeErr != nil {
			fmt.Fprintf(os.Stderr, "warning: close editor: %v\n", closeErr)
		}
	}()

	// Step 3: try a built-in theme preset ("integral").
	// SetGlobalThemePreset rewrites the theme XML embedded in the PPTX.
	// It may return an error when the base presentation has no theme part yet;
	// log it as a warning rather than aborting.
	if err := e.SetGlobalThemePreset("integral"); err != nil {
		log.Printf("Note: SetGlobalThemePreset: %v", err)
	}

	// Step 4: apply a direct styling.Theme (ThemeTech).
	// ApplyTheme rewrites theme colors and fonts from a Go-native Theme struct.
	// Like SetGlobalThemePreset, it requires a theme part; log errors as warnings.
	if err := e.ApplyTheme(styling.ThemeTech); err != nil {
		log.Printf("Note: ApplyTheme(ThemeTech): %v", err)
	}

	// Step 5: add a reference slide listing every available preset and theme var.
	_, err = e.AddSlide(buildReferenceSlide())
	if err != nil {
		return fmt.Errorf("add reference slide: %w", err)
	}

	// Step 6: save.
	outputPath := filepath.Join(outputDir, outputFile)
	if err := e.Save(outputPath); err != nil {
		return fmt.Errorf("save: %w", err)
	}

	log.Printf("Generated %s\n", outputPath)
	return nil
}

// buildBase creates a presentation with a few slides for theme-application demos.
func buildBase(path string) error {
	builder := pptx.NewPresentationBuilder("Advanced Theme Management")

	builder.AddSlide(
		pptx.NewSlide("Theme Management Overview").
			AddBullet("SetGlobalThemePreset – apply an OOXML built-in preset").
			AddBullet("ApplyTheme – apply a Go-native styling.Theme struct").
			AddBullet("Both methods rewrite the embedded theme XML"),
	)

	builder.AddSlide(
		pptx.NewSlide("Available Presets").
			AddNumbered("office2013").
			AddNumbered("facet").
			AddNumbered("integral").
			AddNumbered("ion").
			AddNumbered("retrospect").
			AddNumbered("slice").
			AddNumbered("wisp"),
	)

	builder.AddSlide(
		pptx.NewSlide("Available styling.Theme Variables").
			AddNumbered("styling.ThemeCorporate – professional and trustworthy").
			AddNumbered("styling.ThemeModern   – clean and simple").
			AddNumbered("styling.ThemeTech     – modern technology feel").
			AddNumbered("styling.ThemeDark     – easy on the eyes").
			AddNumbered("styling.ThemeVibrant  – bold and colorful").
			AddNumbered("styling.ThemeNature   – fresh and organic").
			AddNumbered("styling.ThemeCarbon   – IBM Carbon design"),
	)

	return builder.WriteToFile(path)
}

// buildReferenceSlide summarises what was applied so it is visible in the deck.
func buildReferenceSlide() pptx.SlideContent {
	return pptx.NewSlide("Applied in This Run").
		AddBullet("SetGlobalThemePreset(\"integral\") – attempted").
		AddBullet("ApplyTheme(styling.ThemeTech)   – attempted").
		AddBullet("Errors logged as warnings if no theme part present")
}

File: examples/43-advanced-theme-management/main.py
"""Demonstrate advanced theme management with built-in presets.

This example demonstrates:
- Applying a built-in theme preset (ThemeTech)
- Applying the ThemeCorporate and ThemeModern presets
- Adding a reference slide listing all available presets
- Theme color and font slot documentation
"""

from __future__ import annotations

from pathlib import Path

from gopptx import Presentation
from gopptx.presentation.theme import get_theme


def _add_base_content(prs: Presentation) -> None:
    """Add base content slides before applying themes."""
    prs.add_bullet_slide(
        "Advanced Theme Management",
        [
            "Apply built-in themes to existing presentations",
            "SetGlobalThemePreset — rewrites embedded theme XML",
            "ApplyTheme — applies Go-native Theme struct",
            "Themes control: colors, fonts, backgrounds",
        ],
    )
    prs.add_bullet_slide(
        "Theme Workflow",
        [
            "1. Build base presentation with content",
            "2. Open with editor",
            "3. Apply theme preset or custom theme",
            "4. Add reference slides and save",
        ],
    )


def _add_reference_slide(prs: Presentation) -> None:
    """Add a reference slide listing available presets."""
    prs.add_bullet_slide(
        "Available Theme Presets",
        [
            "aurora — cool blues and teals",
            "ocean  — deep blues and greens",
            "sunset — warm oranges and reds",
            "forest — greens and earth tones",
        ],
    )
    prs.add_bullet_slide(
        "Theme Color Slots",
        [
            "Dk1 / Lt1  — primary dark and light (text / background)",
            "Dk2 / Lt2  — secondary dark and light",
            "Accent1-6  — six highlight colors for charts and shapes",
            "Hlink      — hyperlink color",
            "FolHlink   — followed-hyperlink color",
        ],
    )
    prs.add_bullet_slide(
        "Theme Font Slots",
        [
            "MajorFont — heading typeface",
            "MinorFont — body/paragraph typeface",
            "PowerPoint resolves +mj-lt and +mn-lt against these slots",
        ],
    )


def main() -> None:
    """Create presentation demonstrating advanced theme management."""
    output_dir = Path("examples/output")
    output_dir.mkdir(exist_ok=True)

    with Presentation.new("Advanced Theme Management") as prs:
        _add_base_content(prs)

        # Apply a valid built-in theme preset from the Python theme registry.
        prs.apply_theme(get_theme("ocean"))
        theme_inventory = prs.get_theme_inventory()

        _add_reference_slide(prs)

        prs.add_bullet_slide(
            "Theme Inventory Snapshot",
            [
                f"Inventory keys reported: {len(theme_inventory)}",
                "Theme applied via get_theme('ocean')",
                "Reference slides list the available built-in presets",
            ],
        )

        output_path = output_dir / "43_advanced_theme_management.pptx"
        prs.save(str(output_path))
        print(f"Saved: {output_path}")

    print("\n=== SUMMARY ===")
    print("5 slides with the Ocean theme applied")
    print("Covers: presets, color slots, font slots")


if __name__ == "__main__":
    main()
